File: java/learning-spring-02/src/main/java/com/example/learningspring/beans/helloworld/HelloWorldConfiguration.java
package com.example.learningspring.beans.helloworld;

import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Primary;

// Eliminate verbosity in creating Java Beans =>
// Public accessor methods, constructor, equals, hashcode and toString이 자동으로 만들어진다.
// Flutter에서 freeze 느낌인 것 같다.
// JDK 16에서부터 사용가능하다.
record Person(String name, int age, Address address) {
}

record Address(String firstLine, String city) {

}

@Configuration
public class HelloWorldConfiguration {
    @Bean
    public String name() {
        return "Ranga";
    }

    @Bean
    public int age() {
        return 15;
    }

    @Bean
    public Person person() {
        return new Person("Ravi", 20, new Address("Baker Street", "London"));
    }

    @Bean
    public Person person2MethodCall() {
        return new Person(name(), age(), address());
    }

    @Bean
    public Person person3Parameters(String name, int age, Address address3) {
        return new Person(name, age, address3);
    }

    @Bean
    @Primary
    public Person person4Parameters(String name, int age, Address address) {
        return new Person(name, age, address);
    }

    @Bean
    public Person person5Qualifier(String name, int age, @Qualifier("address3qualifier") Address address) {
        return new Person(name, age, address);
    }

    // rename을 하면 기존 이름으로는 사용이 불가능하다.
    @Bean(name = "address2")
    @Primary
    public Address address() {
        return new Address("Baker Street", "London");
    }

    @Bean(name = "address3")
    @Qualifier("address3qualifier")
    public Address address3() {
        return new Address("Motinagar", "Hyderabad");
    }
}

File: kotlin/learning-spring/src/main/kotlin/com/examplekotlin/learningspring/beans/helloworld/HelloWorldConfiguration.kt
package com.examplekotlin.learningspring.beans.helloworld

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

data class Person(val name: String, val age: Int, val address: Address)
data class Address(val firstLine: String, val city: String)

@Configuration
class HelloWorldConfiguration {
    @Bean
    fun name(): String {
        return "Ranga";
    }

    @Bean
    fun age(): Int {
        return 15
    }

    @Bean
    fun person(): Person {
        return Person("Ravi", 20, Address("Baker Street", "London"))
    }

    @Bean
    fun person2MethodCall(): Person {
        return Person(name(), age(), address())
    }

    @Bean
    fun person3Parameters(name: String, age: Int, address3: Address): Person {
        return Person(name, age, address3)
    }

    @Bean("address2")
    fun address(): Address {
        return Address("Baker Street", "London")
    }

    // 자바에서는 String[]를 String처럼 써도 알아서 처리된다.
    // 코틀린에서는 Array라는 명시적 타입이 있어서 안되는 것 같다.
    @Bean(name = ["address3"])
    fun address3(): Address {
        return Address("Motinagar", "Hyderabad")
    }
}